scopesim/server/database.py

```diff
@@ -68,19 +307,135 @@ def list_packages(pkg_name=None):
         list_packages("Armazones")
 
     """
-    pkgs_dict = get_server_package_list()
+    all_grouped = get_all_package_versions()
 
     if pkg_name is None:
-        pkg_names = list(pkgs_dict.keys())
-    elif pkg_name in pkgs_dict:
-        path = pkgs_dict[pkg_name]["path"]
-        pkgs = get_server_folder_contents(path)
-        pkg_names = [pkg for pkg in pkgs if pkg_name in pkg]
+        # Return all packages with any stable version
+        all_stable = list(dict(get_all_stable(all_grouped)).keys())
+        return all_stable
+
+    if not pkg_name in all_grouped:
+        raise ValueError(f"Package name {pkg_name} not found on server.")
+
+    p_versions = [_unparse_raw_version(version, pkg_name)
+                  for version in all_grouped[pkg_name]]
+    return p_versions
+
+
+def _get_zipname(pkg_name: str, release: str, all_versions) -> str:
+    if release == "stable":
+        zip_name = get_stable(all_versions[pkg_name])
+    elif release == "latest":
+        zip_name = get_latest(all_versions[pkg_name])
+    else:
+        release = _parse_raw_version(release)
+        if release not in all_versions[pkg_name]:
+            msg = (f"Requested version '{release}' of '{pkg_name}' package"
+                   " could not be found on the server. Available versions "
+                   f"are: {all_versions[pkg_name]}")
+            raise ValueError(msg)
+        zip_name = release
+    return _unparse_raw_version(zip_name, pkg_name)
+
+
+def _create_session(cached: bool = False, cache_name: str = ""):
+    if cached:
+        return CachedSession(cache_name)
+    return requests.Session()
+
+
+def _initiate_download(pkg_url: str,
+                       cached: bool = False, cache_name: str = "",
+                       total: int = 5, backoff_factor: int = 2):
+    retry_strategy = Retry(total=total, backoff_factor=backoff_factor,
+                           status_forcelist=[429, 500, 501, 502, 503],
+                           allowed_methods=["GET"])
+    adapter = HTTPAdapter(max_retries=retry_strategy)
+    with _create_session(cached, cache_name) as session:
+        session.mount("https://", adapter)
+        response = session.get(pkg_url, stream=True)
+    return response
+
+
+def _handle_download(response, save_path: Path, pkg_name: str,
+                     padlen: int, chunk_size: int = 128) -> None:
+    try:
+        tqdm_kwargs["desc"] = f"Downloading {pkg_name:<{padlen}}"
+        with tqdm.wrapattr(save_path.open("wb"), "write", miniters=1,
+                           total=int(response.headers.get("content-length", 0)),
+                           **tqdm_kwargs) as file:
+            for chunk in response.iter_content(chunk_size=chunk_size):
+                file.write(chunk)
+    except Exception as error:
+        raise error
+    finally:
+        file.close()
```

Comment:
I was surprised by this try/except/finally block, because "it shouldn't do anything due to the context manager". But after some investigation the finally tuns out to be necessary, because `save_path.open()` is not in a context manager.

It seems you could do this:
```suggestion
    tqdm_kwargs["desc"] = f"Downloading {pkg_name:<{padlen}}"
    total=int(response.headers.get("content-length", 0))
    with (
        save_path.open("wb") as file_outer,
        tqdm.wrapattr(file_outer, "write", miniters=1, total=total, **tqdm_kwargs) as file_inner,
    ):
        for chunk in response.iter_content(chunk_size=chunk_size):
            file_inner.write(chunk)
```

which is 1 indentation level less, and 2 lines shorter (or 3 if you move the sad smiley to the previous line), but does behave the same, as both `file_inner` and `file_outer` are closed even if an exception occurs.

Reply:
Testing code, in case you are interested:
```python
from pathlib import Path
from tqdm import tqdm

try:
    save_path = Path('test.txt')
    with (
        save_path.open("w") as file_outer,
        tqdm.wrapattr(file_outer, "write", total=10, desc="hii") as file_inner
    ):
        for ii in range(10):
            file_inner.write(str(ii))
            assert ii < 5

except:
    print(f"{file_outer.closed=}")
    print(f"{file_inner.closed=}")

```

Reply:
Oh yes, the dark art of the double with-block! I originally ran into the problem that the zip file extraction failed, took me a few minutes to figure out the files were not properly released, so I added the `close()`. But your solution is clearly superior, and the way it should be done in Python.

Reply:
It seems Python 3.8 doesn't understand this double with statement in parentheses. But it gets weirder: According to the docs, this is a new feature in Python 3.10. I'm on 3.9.16, but **it works on my machine** 🤓 
Anyway, two *nested* with-blocks should work. Once we can drop support for 3.8 (or maybe 3.9?), this can be changed to the non-nested variant...